catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileEntry.java

```diff
@@ -0,0 +1,242 @@
+ * snapshot will NOT update while in a non-committed state to avoid files "showing" they've been
+ * updated when the changes haven't gone through yet. Once the changes are committed and finalized
+ * then it's up to the user to tell the AsyncFileEntry that the changes were successful by
+ * calling commit().
+ *
+ * @apiNote Once hasChanged() returns true it's up to the user to put the {@link AsyncFileEntry}
+ *     back into a finalized state by calling commit().
+ * @apiNote isInitCommit() will tell if the file has ever been committed. Useful for sending
+ *     additional create operations even if it appears from a child.
+ */
+public class AsyncFileEntry implements Serializable, Comparable<AsyncFileEntry> {
+
+  private final File contentFile;
+  private boolean exists;
+  private long lastModified;
+  private boolean directory;
+  private long length;
+
+  private boolean committed = false;
+
+  private boolean initCommit = false;
+
+  private final Set<AsyncFileEntry> children = new ConcurrentSkipListSet<>();
+
+  private AsyncFileEntry parent;
+
+  private String name;
+
+  public AsyncFileEntry(File file) {
+    this(null, file);
+  }
+
+  public AsyncFileEntry(AsyncFileEntry parent, File file) {
+    this.parent = parent;
+    contentFile = file;
+    refresh();
+  }
+
+  private void refresh() {
+    name = contentFile.getName();
+    exists = contentFile.exists();
+    lastModified = exists ? contentFile.lastModified() : 0;
+    directory = exists && contentFile.isDirectory();
+    length = exists && !directory ? contentFile.length() : 0;
+  }
+
+  public synchronized boolean hasChanged() {
+
+    //  If the file is not committed it "hasChanged" and we don't want to update the snapshot (case
+    // it changes again)
+    if (!committed) {
+      return true;
+    }
+
+    // cache original values
+    final boolean origExists = exists;
+    final long origLastModified = lastModified;
+    final boolean origDirectory = directory;
+    final long origLength = length;
+
+    // refresh the values
+    refresh();
+
+    //  Checks to see if the file has been changed
+    boolean changed =
+        exists != origExists
+            || lastModified != origLastModified
+            || directory != origDirectory
+            || length != origLength;
+
+    if (changed) {
+      committed = false;
+    }
+
+    return changed;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public File getFile() {
+    return contentFile;
+  }
+
+  public boolean isDirectory() {
+    return directory;
+  }
+
+  public Optional<AsyncFileEntry> getParent() {
+    return Optional.ofNullable(parent);
+  }
+
+  /**
+   * checkNetwork:
+   *
+   * <p>Checking the network by checking the directory under the file. This works under two
+   * assumptions:
+   * <ol>
+   *     <li>The root directory will NOT be deleted</li>
+   *     <li>There will NOT be a semlink to another NFS</li>
```

Comment:
📘 If this second assumption can't be made then we can change the logic here to stop when the root directory or a symlink has been hit